kernel/src/collections/list_v.rs

```diff
@@ -0,0 +1,606 @@
+#![feature(const_refs_to_cell)]
+
+use vstd::prelude::*;
+
+verus! {
+
+use vstd::cell::*;
+
+pub struct ListLinkV<'a, T: 'a + ?Sized>(pub PCell<Option<&'a T>>);
+
+impl<'a, T: ?Sized> ListLinkV<'a, T> {
+    pub const fn empty() -> (res: (ListLinkV<'a, T>, Tracked<PointsTo<Option<&'a T>>>))
+        ensures
+            res.0.0.id() == res.1@.id(),
+            res.1@.is_init(),
+            match res.1@.value() {
+                Option::Some(_) => false,
+                Option::None => true,
+            },
+    {
+        let (cell, points_to) = PCell::<Option<&'a T>>::new(None);
+        (ListLinkV(cell), points_to)
+    }
+}
+
+pub trait ListNodeV<'a, T: ?Sized> {
+    fn next(&'a self, next_points_to: Tracked<&PointsTo<Option<&'a T>>>) -> (res: &'a ListLinkV<
+        'a,
+        T,
+    >)
+        ensures
+            res.0.id() == next_points_to@.id(),
+    ;
+}
+
+pub struct ListV<'a, T: 'a + ?Sized + ListNodeV<'a, T>> {
+    pub head: ListLinkV<'a, T>,
+}
+
+pub tracked struct GhostState<'a, T: 'a + ?Sized + ListNodeV<'a, T>> {
+    pub ghost cells: Seq<&'a PCell<Option<&'a T>>>,
+    pub tracked points_to_map: Map<nat, PointsTo<Option<&'a T>>>,
+}
+
+pub struct ListIteratorV<'a, T: 'a + ?Sized + ListNodeV<'a, T>> {
+    pub cur: Option<&'a T>,
+    pub index: Ghost<nat>,
+}
+
+pub closed spec fn well_formed_node<'a, T: 'a + ?Sized + ListNodeV<'a, T>>(
+    ghost_state: &Tracked<GhostState<'a, T>>,
+    i: nat,
+) -> bool {
+    &&& ghost_state@.points_to_map.dom().contains(i)
+    &&& ghost_state@.points_to_map[i].is_init()
+    &&& ghost_state@.points_to_map[i].id() == ghost_state@.cells[i as int].id()
+}
+
+impl<'a, T: ?Sized + ListNodeV<'a, T>> ListIteratorV<'a, T> {
+    pub fn new(l: &'a ListV<'a, T>, ghost_state: &Tracked<GhostState<'a, T>>) -> (res:
+        ListIteratorV<'a, T>)
+        requires
+            ghost_state@.cells.len() >= 1,
+            forall|i: nat|
+                0 <= i < ghost_state@.cells.len()
+                    ==> #[trigger] ghost_state@.points_to_map.dom().contains(i)
+                    && ghost_state@.points_to_map[i].is_init() && ghost_state@.points_to_map[i].id()
+                    == ghost_state@.cells[i as int].id(),
+            forall|i: nat|
+                0 <= i < (ghost_state@.cells.len() - 1) as nat
+                    ==> match #[trigger] ghost_state@.points_to_map[i].value() {
+                    Option::Some(_) => true,
+                    Option::None => false,
+                },
+            match ghost_state@.points_to_map[(ghost_state@.cells.len() - 1) as nat].value() {
+                Option::Some(_) => false,
+                Option::None => true,
+            },
+            ghost_state@.cells[0].id() == l.head.0.id(),
+        ensures
```

Comment:
They are not exactly the same, and I do not think the one here can be put in the same spec function as those in `next` and `last`, because the latter ones involves using `self`. The latter ones may indeed be put into the same spec function.